pkg/service/service_test.go

```diff
@@ -1469,23 +1468,34 @@ func (m *ManagerTestSuite) TestL7LoadBalancerServiceOverride(c *C) {
 	// registering without redirecting
 	echoOtherNode := lb.ServiceName{Name: "echo-other-node", Namespace: "cilium-test"}
 	resource1 := lb.ServiceName{Name: "testOwner1", Namespace: "cilium-test"}
-	err = m.svc.RegisterL7LBServiceBackendSync(echoOtherNode, resource1, nil)
+	err = m.svc.RegisterL7LBService(echoOtherNode, resource1, nil, 0)
+	c.Assert(err, IsNil)
+
+	svc, ok = m.svc.svcByID[id]
+	c.Assert(len(svc.backends), Equals, len(allBackends))
+	c.Assert(ok, Equals, true)
+	c.Assert(svc.l7LBProxyPort, Equals, uint16(0))
+
+	// registering backend sync for resource1
+	err = m.svc.RegisterL7LBServiceBackendSync(echoOtherNode, resource1, []string{"8080", "9090", "10000"})
 	c.Assert(err, IsNil)
 
 	svc, ok = m.svc.svcByID[id]
 	c.Assert(len(svc.backends), Equals, len(allBackends))
 	c.Assert(ok, Equals, true)
 	c.Assert(svc.l7LBProxyPort, Equals, uint16(0))
+	c.Assert(svc.l7LBFrontendPorts, IsNil) // No forward active
 
 	// registering with redirection stores the proxy port
 	resource2 := lb.ServiceName{Name: "testOwner2", Namespace: "cilium-test"}
-	err = m.svc.RegisterL7LBService(echoOtherNode, resource2, nil, uint16(9090))
+	err = m.svc.RegisterL7LBService(echoOtherNode, resource2, []string{"8080", "9090", "11000"}, uint16(9090))
 	c.Assert(err, IsNil)
 
 	svc, ok = m.svc.svcByID[id]
 	c.Assert(len(svc.backends), Equals, len(allBackends))
 	c.Assert(ok, Equals, true)
 	c.Assert(svc.l7LBProxyPort, Equals, uint16(9090))
+	c.Assert(svc.l7LBFrontendPorts, checker.DeepEquals, []string{"10000", "11000", "8080", "9090"})
```

Comment:
The PR commit message mentions "endpoints not being synchronized".  Is there a better way to test the expected behavior in terms of black box testing other than asserting on some internal field in the service struct that could potentially change in future?

Reply:
This should be possible by setting up the servicemanager (under test) with a mock for the `envoyCache` (xDS server interface within the ServiceManager) and assert what's actually passed to it.

But with the plan to refactor this quite a bit by extracting the Envoy logic from the service package with https://github.com/cilium/cilium/pull/30184: Would it be OK for your if we improve testing based on the refactored state instead of this bugfix PR? Because after this refactoring, the ServiceManager would no longer have a dependency to Envoy at all.

Reply:
Sure, fine by me.